core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java

```diff
@@ -0,0 +1,745 @@
+        }
+
+        @Override
+        public boolean run(int workerId) {
+            assert workerId == id;
+            try {
+                long cursor;
+                while ((cursor = sequence.next()) < 0) {
+                    if (cursor == -1) {
+                        return false;
+                    }
+                }
+                LineTcpMeasurementEvent event = queue.get(cursor);
+                boolean eventProcessed;
+                try {
+                    if (event.threadId == id) {
+                        eventProcessed = processNextEvent(event);
+                    } else {
+                        if (event.isRebalanceEvent()) {
+                            eventProcessed = processRebalance(event);
+                        } else {
+                            eventProcessed = true;
+                        }
+                    }
+                } catch (RuntimeException ex) {
+                    LOG.error().$(ex).$();
+                    eventProcessed = true;
+                }
+                if (eventProcessed) {
+                    sequence.done(cursor);
+                }
+            } catch (RuntimeException ex) {
+                LOG.error().$(ex).$();
+            }
+            return true;
+        }
+
+        private boolean processNextEvent(LineTcpMeasurementEvent event) {
+            Parser parser = parserCache.get(event.getTableName());
+            if (null == parser) {
+                parser = new Parser();
+                try {
+                    parser.processFirstEvent(engine, securityContext, event);
+                } catch (CairoException ex) {
+                    LOG.info().$(name).$(" could not create parser [name=").$(event.getTableName()).$(", ex=").$(ex.getFlyweightMessage()).$(']').$();
+                    parser.close();
+                    return false;
+                }
+                LOG.info().$(name).$(" created parser [name=").$(event.getTableName()).$(']').$();
+                parserCache.put(event.getTableName().toString(), parser);
+                return true;
+            } else {
+                parser.processEvent(event);
+                return true;
+            }
+        }
+
+        private boolean processRebalance(LineTcpMeasurementEvent event) {
+            if (event.rebalanceToThreadId == id) {
+                if (!event.rebalanceReleasedByFromThread) {
+                    return false;
+                }
+
+                return true;
+            }
+
+            if (event.rebalanceFromThreadId == id) {
+                Parser parser = parserCache.get(event.rebalanceTableName);
+                parserCache.remove(event.rebalanceTableName);
+                parser.close();
+                event.rebalanceReleasedByFromThread = true;
+                return true;
+            }
+
+            return true;
+        }
+
+        private class Parser implements Closeable {
+            private TableWriter writer;
```

Comment:
it looks like parser can boil down to the map of table writers. Parser state is transient for the row of data and can be part of the WriterJob